src/QueryBuilder.php

```diff
@@ -215,9 +215,22 @@ private function includeAccessData(string $index): bool
             return false;
         }
 
-        $permissionProvider = $this->getServiceLocator()->get(PermissionInterface::SERVICE_ID);
+        $permissionProvider = $this->getPermissionProvider();
+
+        if (!$permissionProvider instanceof ReverseRightLookupInterface) {
+            return false;
+        }
+
+        $currentUser = $this->getSessionService()->getCurrentUser();
+
+        $nonExistingId = uniqid();
+        $permissions = $permissionProvider->getPermissions($currentUser, [$nonExistingId])[$nonExistingId] ?? [];
 
-        return $permissionProvider instanceof ReverseRightLookupInterface;
+        if (in_array(PermissionInterface::RIGHT_READ, $permissions)) {
+            return false;
+        }
+
+        return true;
```

Comment:
```suggestion
        return !in_array(PermissionInterface::RIGHT_READ, $permissions, true);
```

Reply:
Done, fixed also the previous code and moved to an specification.